// mockttp/dist/admin/mockttp-admin-plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockttpAdminPlugin = void 0;
const mockttp_server_1 = require("../server/mockttp-server");
const mockttp_admin_model_1 = require("./mockttp-admin-model");
const mockttp_schema_1 = require("./mockttp-schema");
class MockttpAdminPlugin {
    mockServer;
    messageBodyDecoding;
    async start({ port, options, messageBodyDecoding }) {
        this.mockServer = new mockttp_server_1.MockttpServer(options);
        this.messageBodyDecoding = messageBodyDecoding ||
            'none'; // Backward compat - clients that don't set this option expect 'none'.
        await this.mockServer.start(port);
        return {
            port: this.mockServer.port,
            mockRoot: this.mockServer.url
        };
    }
    stop() {
        return this.mockServer.stop();
    }
    reset() {
        return this.mockServer.reset();
    }
    getMockServer() {
        return this.mockServer;
    }
    enableDebug() {
        this.mockServer.enableDebug();
    }
    schema = mockttp_schema_1.MockttpSchema;
    buildResolvers(stream, ruleParameters) {
        return (0, mockttp_admin_model_1.buildAdminServerModel)(this.mockServer, stream, ruleParameters, {
            messageBodyDecoding: this.messageBodyDecoding
        });
    }
    ;
}
exports.MockttpAdminPlugin = MockttpAdminPlugin;
