// mockttp/dist/rules/rule-deserialization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeRuleData = deserializeRuleData;
exports.deserializeWebSocketRuleData = deserializeWebSocketRuleData;
const serialization_1 = require("../serialization/serialization");
const matchers = require("./matchers");
const completionCheckers = require("./completion-checkers");
const request_step_impls_1 = require("./requests/request-step-impls");
const websocket_step_impls_1 = require("./websockets/websocket-step-impls");
function deserializeRuleData(data, stream, options) {
    const steps = data.steps
        ? data.steps
        : data.handler
            ? [data.handler]
            : [];
    return {
        id: data.id,
        priority: data.priority,
        matchers: data.matchers.map((m) => (0, serialization_1.deserialize)(m, stream, options, matchers.MatcherLookup)),
        steps: steps.map(step => (0, serialization_1.deserialize)(step, stream, options, request_step_impls_1.StepLookup)),
        completionChecker: data.completionChecker && (0, serialization_1.deserialize)(data.completionChecker, stream, options, completionCheckers.CompletionCheckerLookup)
    };
}
function deserializeWebSocketRuleData(data, stream, options) {
    const steps = data.steps
        ? data.steps
        : data.handler
            ? [data.handler]
            : [];
    return {
        id: data.id,
        matchers: data.matchers.map((m) => (0, serialization_1.deserialize)(m, stream, options, matchers.MatcherLookup)),
        steps: steps.map(step => (0, serialization_1.deserialize)(step, stream, options, websocket_step_impls_1.WsStepLookup)),
        completionChecker: data.completionChecker && (0, serialization_1.deserialize)(data.completionChecker, stream, options, completionCheckers.CompletionCheckerLookup)
    };
}
