// read-tls-client-hello/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NonTlsError = exports.isGREASE = exports.extensionParsers = void 0;
exports.getExtensionData = getExtensionData;
exports.readTlsClientHello = readTlsClientHello;
exports.calculateJa3 = calculateJa3;
exports.getTlsFingerprintAsJa3 = getTlsFingerprintAsJa3;
exports.calculateJa4 = calculateJa4;
exports.getTlsFingerprintAsJa4 = getTlsFingerprintAsJa4;
exports.trackClientHellos = trackClientHellos;
const stream = require("stream");
const crypto = require("crypto");
const extension_parsers_1 = require("./extension-parsers");
const lookup_tables_1 = require("./lookup-tables");
var extension_parsers_2 = require("./extension-parsers");
Object.defineProperty(exports, "extensionParsers", { enumerable: true, get: function () { return extension_parsers_2.extensionParsers; } });
__exportStar(require("./lookup-tables"), exports);
const collectBytes = (stream, byteLength) => {
    if (byteLength === 0)
        return Buffer.from([]);
    return new Promise((resolve, reject) => __awaiter(void 0, void 0, void 0, function* () {
        var _a;
        const closeReject = () => reject(new Error('Stream closed before expected data could be read'));
        const data = [];
        try {
            stream.on('error', reject);
            stream.on('close', closeReject);
            let dataLength = 0;
            let readNull = false;
            do {
                if (!stream.readable || readNull)
                    yield new Promise((resolve) => stream.once('readable', resolve));
                const nextData = (_a = stream.read(byteLength - dataLength)) !== null && _a !== void 0 ? _a : stream.read(); // If less than wanted data is available, at least read what we can get
                if (nextData === null) {
                    // Still null => tried to read, not enough data
                    readNull = true;
                    continue;
                }
                data.push(nextData);
                dataLength += nextData.byteLength;
            } while (dataLength < byteLength);
            return resolve(Buffer.concat(data, byteLength));
        }
        catch (e) {
            Object.assign(e, { consumedData: data });
            reject(e);
        }
        finally {
            stream.removeListener('error', reject);
            stream.removeListener('close', closeReject);
        }
    }));
};
// https://datatracker.ietf.org/doc/html/draft-davidben-tls-grease-01 defines GREASE values for various
// TLS fields, reserving 0a0a, 1a1a, 2a2a, etc for ciphers, extension ids & supported groups.
const isGREASE = (value) => (value & 0x0f0f) == 0x0a0a;
exports.isGREASE = isGREASE;
function getExtensionData(clientHello, id) {
    var _a;
    const numId = typeof id === 'string'
        ? lookup_tables_1.EXTENSION_IDS[id]
        : id;
    return (_a = clientHello.extensions.find(e => e.id === numId)) === null || _a === void 0 ? void 0 : _a.data;
}
/**
 * Separate error class. If you want to detect TLS parsing errors, but ignore TLS fingerprint
 * issues from definitely-not-TLS traffic, you can ignore all instances of this error.
 */
class NonTlsError extends Error {
    constructor(message) {
        super(message);
        // Fix prototypes (required for custom error types):
        const actualProto = new.target.prototype;
        Object.setPrototypeOf(this, actualProto);
    }
}
exports.NonTlsError = NonTlsError;
function extractTlsHello(inputStream) {
    return __awaiter(this, void 0, void 0, function* () {
        const consumedData = [];
        try {
            consumedData.push(yield collectBytes(inputStream, 1));
            const [recordType] = consumedData[0];
            if (recordType !== 0x16)
                throw new Error("Can't calculate TLS fingerprint - not a TLS stream");
            consumedData.push(yield collectBytes(inputStream, 2));
            const recordLengthBytes = yield collectBytes(inputStream, 2);
            consumedData.push(recordLengthBytes);
            const recordLength = recordLengthBytes.readUInt16BE(0);
            consumedData.push(yield collectBytes(inputStream, recordLength));
            // Put all the bytes back, so that this stream can still be used to create a real TLS session
            return Buffer.concat(consumedData);
        }
        catch (error) {
            if (error.consumedData) {
                // This happens if there's an error inside collectBytes with a partial read.
                error.consumedData.consumedData = Buffer.concat([
                    ...consumedData,
                    error.consumedData
                ]);
            }
            else {
                Object.assign(error, { consumedData: Buffer.concat(consumedData) });
            }
            throw error;
        }
    });
}
function readTlsClientHello(inputStream) {
    return __awaiter(this, void 0, void 0, function* () {
        const wasFlowing = inputStream.readableFlowing;
        if (wasFlowing)
            inputStream.pause(); // Pause other readers, so we have time to precisely get the data we need.
        let clientHelloRecordData;
        try {
            clientHelloRecordData = yield extractTlsHello(inputStream);
        }
        catch (error) {
            if ('consumedData' in error) {
                inputStream.unshift(error.consumedData);
            }
            if (wasFlowing)
                inputStream.resume(); // If there were other readers, resume and let them continue
            throw new NonTlsError(error.message);
        }
        // Put all the bytes back, so that this stream can still be used to create a real TLS session
        inputStream.unshift(clientHelloRecordData);
        if (wasFlowing)
            inputStream.resume(); // If there were other readers, resume and let them continue
        // Collect all the hello bytes, and then give us a stream of exactly only those bytes, so we can
        // still process them step by step in order:
        const clientHello = clientHelloRecordData.slice(5); // Strip TLS record prefix
        const helloDataStream = stream.Readable.from(clientHello, { objectMode: false });
        const [helloType] = (yield collectBytes(helloDataStream, 1));
        if (helloType !== 0x1)
            throw new Error("Can't calculate TLS fingerprint - not a TLS client hello");
        const helloLength = (yield collectBytes(helloDataStream, 3)).readIntBE(0, 3);
        if (helloLength !== clientHello.byteLength - 4)
            throw new Error(`Unexpected client hello length: ${helloLength} (of ${clientHello.byteLength})`);
        const clientTlsVersion = yield collectBytes(helloDataStream, 2);
        const clientRandom = yield collectBytes(helloDataStream, 32);
        const [sessionIdLength] = yield collectBytes(helloDataStream, 1);
        const sessionId = yield collectBytes(helloDataStream, sessionIdLength);
        const cipherSuitesLength = (yield collectBytes(helloDataStream, 2)).readUInt16BE(0);
        const cipherSuites = yield collectBytes(helloDataStream, cipherSuitesLength);
        const [compressionMethodsLength] = yield collectBytes(helloDataStream, 1);
        const compressionMethods = yield collectBytes(helloDataStream, compressionMethodsLength);
        const allCipherSuiteIds = [];
        for (let i = 0; i < cipherSuites.length; i += 2) {
            allCipherSuiteIds.push(cipherSuites.readUInt16BE(i));
        }
        const allCompressionMethods = Array.from(compressionMethods);
        const extensionsLength = (yield collectBytes(helloDataStream, 2)).readUInt16BE(0);
        let readExtensionsDataLength = 0;
        const parsedExtensions = [];
        while (readExtensionsDataLength < extensionsLength) {
            const extensionId = (yield collectBytes(helloDataStream, 2)).readUInt16BE(0);
            const extensionLength = (yield collectBytes(helloDataStream, 2)).readUInt16BE(0);
            const extensionData = yield collectBytes(helloDataStream, extensionLength);
            let parsedData = null;
            const parser = extension_parsers_1.extensionParsers[extensionId];
            if (parser && !(0, exports.isGREASE)(extensionId)) {
                try {
                    parsedData = parser(extensionData);
                }
                catch (_a) {
                    parsedData = null; // Malformed extension data - fall back gracefully
                }
            }
            parsedExtensions.push({ id: extensionId, data: parsedData });
            readExtensionsDataLength += 4 + extensionLength;
        }
        return {
            version: clientTlsVersion.readUInt16BE(0),
            random: clientRandom,
            sessionId,
            cipherSuites: allCipherSuiteIds,
            compressionMethods: allCompressionMethods,
            extensions: parsedExtensions
        };
    });
}
function calculateJa3(clientHello) {
    var _a, _b, _c, _d;
    const ciphers = clientHello.cipherSuites.filter(id => !(0, exports.isGREASE)(id));
    const extensionIds = clientHello.extensions.map(e => e.id).filter(id => !(0, exports.isGREASE)(id));
    const groups = (_b = (_a = getExtensionData(clientHello, 0x000A)) === null || _a === void 0 ? void 0 : _a.groups.filter(id => !(0, exports.isGREASE)(id))) !== null && _b !== void 0 ? _b : [];
    const curveFormats = (_d = (_c = getExtensionData(clientHello, 0x000B)) === null || _c === void 0 ? void 0 : _c.formats) !== null && _d !== void 0 ? _d : [];
    const fingerprintString = [
        clientHello.version,
        ciphers.join('-'),
        extensionIds.join('-'),
        groups.join('-'),
        curveFormats.join('-')
    ].join(',');
    return crypto.createHash('md5').update(fingerprintString).digest('hex');
}
function getTlsFingerprintAsJa3(rawStream) {
    return __awaiter(this, void 0, void 0, function* () {
        return calculateJa3(yield readTlsClientHello(rawStream));
    });
}
function calculateJa4(clientHello) {
    var _a, _b, _c, _d, _e;
    const ciphers = clientHello.cipherSuites.filter(id => !(0, exports.isGREASE)(id));
    const extensionIds = clientHello.extensions.map(e => e.id).filter(id => !(0, exports.isGREASE)(id));
    const serverName = (_a = getExtensionData(clientHello, 0x0000)) === null || _a === void 0 ? void 0 : _a.serverName;
    const alpnProtocols = (_b = getExtensionData(clientHello, 0x0010)) === null || _b === void 0 ? void 0 : _b.protocols;
    const sigAlgorithms = (_d = (_c = getExtensionData(clientHello, 0x000D)) === null || _c === void 0 ? void 0 : _c.algorithms) !== null && _d !== void 0 ? _d : [];
    // Part A: Protocol info
    const protocol = 't'; // We only handle TCP for now
    const version = extensionIds.includes(0x002B)
        ? '13' // TLS 1.3 uses the supported versions extension, and 1.4+ doesn't exist (yet)
        : (_e = {
            0x0303: '12',
            0x0302: '11',
            0x0301: '10'
        }[clientHello.version]) !== null && _e !== void 0 ? _e : '00'; // Other unknown version
    const sni = !serverName ? 'i' : 'd'; // 'i' for IP (no SNI), 'd' for domain
    // Handle different ALPN protocols
    let alpn = '00';
    const firstProtocol = alpnProtocols === null || alpnProtocols === void 0 ? void 0 : alpnProtocols[0];
    if (firstProtocol && firstProtocol.length >= 1) {
        // Take first and last character of the protocol string
        alpn = firstProtocol.length >= 2
            ? `${firstProtocol[0]}${firstProtocol[firstProtocol.length - 1]}`
            : `${firstProtocol[0]}${firstProtocol[0]}`;
    }
    // Format numbers as fixed-width hex
    const cipherCount = ciphers.length.toString().padStart(2, '0');
    const extensionCount = extensionIds.length.toString().padStart(2, '0');
    const ja4_a = `${protocol}${version}${sni}${cipherCount}${extensionCount}${alpn}`;
    // Part B: Truncated SHA256 of cipher suites
    const cipherHexValues = ciphers
        .map(c => c.toString(16).padStart(4, '0'));
    const sortedCiphers = [...cipherHexValues].sort().join(',');
    const cipherHash = ciphers.length
        ? crypto.createHash('sha256')
            .update(sortedCiphers)
            .digest('hex')
            .slice(0, 12)
        : '000000000000'; // No ciphers provided
    // Part C: Truncated SHA256 of extensions + sig algorithms
    // Get extensions (excluding SNI and ALPN)
    const extensionsStr = extensionIds
        .filter(e => e !== 0x0 && e !== 0x10)
        .sort((a, b) => a - b)
        .map(e => e.toString(16).padStart(4, '0'))
        .join(',');
    const signatureAlgorithmsStr = sigAlgorithms
        .filter(s => !(0, exports.isGREASE)(s))
        .map(s => s.toString(16).padStart(4, '0'))
        .join(',');
    // Add separator only if we have signature algorithms
    const separator = signatureAlgorithmsStr ? '_' : '';
    // Combine and hash
    const ja4_c_raw = `${extensionsStr}${separator}${signatureAlgorithmsStr}`;
    const extensionHash = crypto.createHash('sha256')
        .update(ja4_c_raw)
        .digest('hex')
        .slice(0, 12);
    return `${ja4_a}_${cipherHash}_${extensionHash}`;
}
function getTlsFingerprintAsJa4(rawStream) {
    return __awaiter(this, void 0, void 0, function* () {
        return calculateJa4(yield readTlsClientHello(rawStream));
    });
}
/**
 * Modify a TLS server, so that the TLS client hello is always parsed and the result is
 * attached to all sockets at the point when the 'secureConnection' event fires.
 *
 * This method mutates and returns the TLS server provided. TLS client hello data is
 * available from all TLS sockets afterwards in the `socket.tlsClientHello` property.
 *
 * This will work for all standard uses of a TLS server or similar (e.g. an HTTPS server)
 * but may behave unpredictably for advanced use cases, e.g. if you are already
 * manually injecting connections, hooking methods or events or otherwise doing something
 * funky & complicated. In those cases you probably want to use the fingerprint
 * calculation methods directly inside your funky logic instead.
 */
function trackClientHellos(tlsServer) {
    // Disable the normal TLS 'connection' event listener that triggers TLS setup:
    const tlsConnectionListener = tlsServer.listeners('connection')[0];
    if (!tlsConnectionListener)
        throw new Error('TLS server is not listening for connection events');
    tlsServer.removeListener('connection', tlsConnectionListener);
    // Listen ourselves for connections, get the fingerprint first, then let TLS setup resume:
    tlsServer.on('connection', (socket) => __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        try {
            const clientHello = yield readTlsClientHello(socket);
            socket.tlsClientHello = Object.assign(Object.assign({}, clientHello), { ja3: calculateJa3(clientHello), ja4: calculateJa4(clientHello) });
        }
        catch (e) {
            if (!(e instanceof NonTlsError)) { // Ignore totally non-TLS traffic
                console.warn(`TLS client hello data not available for TLS connection from ${(_a = socket.remoteAddress) !== null && _a !== void 0 ? _a : 'unknown address'}: ${(_b = e.message) !== null && _b !== void 0 ? _b : e}`);
            }
        }
        // Once we have a fingerprint, TLS handshakes can continue as normal:
        tlsConnectionListener.call(tlsServer, socket);
    }));
    tlsServer.prependListener('secureConnection', (tlsSocket) => {
        var _a;
        const fingerprint = (_a = tlsSocket._parent) === null || _a === void 0 ? void 0 : _a.tlsClientHello;
        tlsSocket.tlsClientHello = fingerprint;
    });
    return tlsServer;
}
