// mockttp/dist/rules/requests/request-rule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestRule = void 0;
exports.explainSteps = explainSteps;
const buffer_1 = require("buffer");
const types_1 = require("../../types");
const request_utils_1 = require("../../util/request-utils");
const util_1 = require("@httptoolkit/util");
const matchers = require("../matchers");
const request_step_impls_1 = require("./request-step-impls");
const rule_serialization_1 = require("../rule-serialization");
class RequestRule {
    matchers;
    steps;
    completionChecker;
    id;
    priority;
    requests = [];
    requestCount = 0;
    constructor(data) {
        (0, rule_serialization_1.validateMockRuleData)(data);
        this.id = data.id || crypto.randomUUID();
        this.priority = data.priority ?? types_1.RulePriority.DEFAULT;
        this.matchers = data.matchers;
        this.completionChecker = data.completionChecker;
        this.steps = data.steps.map((stepDefinition, i) => {
            const StepImplClass = request_step_impls_1.StepLookup[stepDefinition.type];
            if (StepImplClass.isFinal && i !== data.steps.length - 1) {
                throw new Error(`Cannot create a rule with a final step before the last position ("${stepDefinition.explain()}" in position ${i + 1} of ${data.steps.length})`);
            }
            // All step impls have a fromDefinition static method that turns a definition into a
            // full impl (copying data and initializing any impl-only fields). Note that for remote clients,
            // the definition itself has already been deserialized by impl.deserialize(data) beforehand.
            return StepImplClass.fromDefinition(stepDefinition);
        });
    }
    matches(request) {
        return matchers.matchesAll(request, this.matchers);
    }
    handle(req, res, options) {
        let stepsPromise = (async () => {
            for (let step of this.steps) {
                const result = await step.handle(req, res, {
                    emitEventCallback: options.emitEventCallback,
                    keyLogStream: options.keyLogStream,
                    debug: options.debug,
                    reqBodyObserved: options.reqBodyObserved
                });
                if (!result || result.continue === false)
                    break;
            }
        })();
        // Requests are added to rule.requests as soon as they start being handled,
        // as promises, which resolve only when the response & request body is complete.
        if (options.record) {
            this.requests.push(Promise.race([
                // When the steps all resolve, the request is completed:
                stepsPromise,
                // If the response is closed before the step completes (due to aborts, step
                // timeouts, whatever) then that also counts as the request being completed:
                new Promise((resolve) => res.on('close', resolve))
            ])
                .catch(() => { }) // Ignore step errors here - we're only tracking the request
                .then(() => (0, request_utils_1.waitForCompletedRequest)(req))
                .catch(() => {
                // If for some reason the request is not completed, we still want to record it.
                // TODO: Update the body to return the data that has been received so far.
                const initiatedRequest = (0, request_utils_1.buildInitiatedRequest)(req);
                return {
                    ...initiatedRequest,
                    body: (0, request_utils_1.buildBodyReader)(buffer_1.Buffer.from([]), req.headers),
                    rawTrailers: [],
                    trailers: {}
                };
            }));
        }
        // Even if traffic recording is disabled, the number of matched
        // requests is still tracked
        this.requestCount += 1;
        return stepsPromise;
    }
    isComplete() {
        if (this.completionChecker) {
            // If we have a specific rule, use that
            return this.completionChecker.isComplete(this.requestCount);
        }
        else if (this.requestCount === 0) {
            // Otherwise, by default we're definitely incomplete if we've seen no requests
            return false;
        }
        else {
            // And we're _maybe_ complete if we've seen at least one request. In reality, we're incomplete
            // but we should be used anyway if we're at any point we're the last matching rule for a request.
            return null;
        }
    }
    explain(withoutExactCompletion = false) {
        let explanation = `Match ${this.priority === types_1.RulePriority.FALLBACK ? 'otherwise unmatched ' : ''}requests ${matchers.explainMatchers(this.matchers)}, and ${explainSteps(this.steps)}`;
        if (this.completionChecker) {
            explanation += `, ${this.completionChecker.explain(withoutExactCompletion ? undefined : this.requestCount)}.`;
        }
        else {
            explanation += '.';
        }
        return explanation;
    }
    dispose() {
        this.steps.forEach(s => s.dispose());
        this.matchers.forEach(m => m.dispose());
        if (this.completionChecker)
            this.completionChecker.dispose();
    }
}
exports.RequestRule = RequestRule;
function explainSteps(steps) {
    return (0, util_1.joinAnd)(steps.map(s => s.explain()), {
        finalSeparator: 'then ',
        oxfordComma: true
    });
}
